brion/synapse.cpp

```diff
@@ -338,47 +350,73 @@ class Synapse : public boost::noncopyable
 
     std::string _findFilename( const uint32_t gid ) const
     {
-        while( _fileMap[ gid ].empty( ))
+        if( !_fileMap[ gid ].empty( ))
+            return _fileMap[ gid ];
+
+        // extract the GID->file mapping from the merge_nrn.sh script
+        const std::string& firstFile = _unmappedFiles.begin()->first;
+        const fs::path filePath = fs::path( firstFile ).parent_path();
+        const fs::path merge_nrn = filePath / "merge_nrn.sh";
+        const std::ifstream mergeFile( merge_nrn.generic_string().c_str( ));
+        if( mergeFile.is_open( ))
         {
-            if( _unmappedFiles.empty( ))
-                return std::string();
+            std::stringstream buffer;
+            buffer << mergeFile.rdbuf();
+
+            const boost::regex commentregx( "#.*?\\n" );
+            const std::string content = boost::regex_replace( buffer.str(),
+                                                           commentregx , "\n" );
+
+            const boost::regex regx( "\\$CMD -i \\$H5.(?<number>[0-9]+) -o "
+                                     "\\$H5 -s /a(?<gid>[0-9]+)" );
+            const int subs[] = {1, 2};
+            boost::sregex_token_iterator i( content.begin(), content.end(),
+                                            regx, subs );
+
+            const std::string basename =
+                       firstFile.substr( 0, firstFile.find_last_of( '.' ) + 1 );
+            for( boost::sregex_token_iterator j; i != j; )
+            {
+                const std::string& fileNumber = *i++;
+                const uint32_t fileGID = boost::lexical_cast< uint32_t >(*i++);
+                _fileMap[ fileGID ] = basename + fileNumber;
+            }
```

Comment:
I'd move the parsing of the merge script to a separate function and call it from the constructor to cache the result because the filesystem is already accessed there to find the candidate files.